boa_gc/src/pointers/weak_map.rs

```diff
@@ -0,0 +1,35 @@
+use crate::{Allocator, Finalize, Gc, GcRefCell, Trace, WeakGc};
+use std::collections::HashMap;
+
+/// A map that holds weak references to its keys and is traced by the garbage collector.
+#[derive(Clone, Debug, Default, Trace, Finalize)]
+pub struct WeakMap<K: Trace + Sized + 'static, V: Trace + Sized + 'static> {
+    pub(crate) inner: Gc<GcRefCell<HashMap<WeakGc<K>, V>>>,
```

Comment:
While this works, I think it would be better to directly use a `HashSet<Ephemeron<K, V>>` to optimize this, since right now it is using an `Ephemeron<K, ()>` and a `V`.

Reply:
Hm, realized that to do this we'll probably have to implement our own `HashMap`...

Reply:
Yeah, this should work for now. Maybe in the future we'll have to inline `HashMap` to optimize the memory of this.